File: core/engine.py
import asyncio
import aiohttp
from core.rpc_manager import RPCManager


class Engine:
    def __init__(self, checker_class, rpc_file, input_file, active_file, inactive_file, chain):
        self.checker_class = checker_class
        self.rpc_file = rpc_file
        self.input_file = input_file
        self.active_file = active_file
        self.inactive_file = inactive_file
        self.chain = chain

        self.addresses = self.load_addresses()
        self.processed = 0
        self.active = 0
        self.inactive = 0
        self.retries = 0

        self.rpc_manager = self.build_rpc_manager()
        self.checker = checker_class(self.rpc_manager)

    def load_addresses(self):
        try:
            with open(self.input_file, "r") as f:
                return [line.strip() for line in f if line.strip()]
        except:
            return []

    def build_rpc_manager(self):
        entries = []

        try:
            with open(self.rpc_file, "r") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue

                    if "|" in line:
                        mode, url = line.split("|", 1)
                        entries.append({
                            "mode": mode.strip(),
                            "url": url.strip()
                        })
        except:
            pass

        return RPCManager(entries)

    async def worker(self, session, address):
        try:
            result = await self.checker.check(session, address)

            self.processed += 1

            if result:
                self.active += 1
                with open(self.active_file, "a") as f:
                    f.write(address + "\n")
            else:
                self.inactive += 1
                with open(self.inactive_file, "a") as f:
                    f.write(address + "\n")

        except:
            self.retries += 1

    async def run(self):
        connector = aiohttp.TCPConnector(limit=100)
        timeout = aiohttp.ClientTimeout(total=15)

        async with aiohttp.ClientSession(connector=connector, timeout=timeout) as session:
            tasks = []

            for address in self.addresses:
                tasks.append(self.worker(session, address))

                if len(tasks) >= 100:
                    await asyncio.gather(*tasks)
                    tasks = []

            if tasks:
                await asyncio.gather(*tasks)

    def status(self):
        return {
            "chain": self.chain,
            "processed": self.processed,
            "active": self.active,
            "inactive": self.inactive,
            "retries": self.retries
        }

File: core/engine.js
const fs = require("fs");
const path = require("path");
const { RPCManager } = require("./rpcManager");

class Engine {
  constructor(CheckerClass, rpcFile, inputFile, activeFile, inactiveFile, chain) {
    this.CheckerClass = CheckerClass;
    this.rpcFile = rpcFile;
    this.inputFile = inputFile;
    this.activeFile = activeFile;
    this.inactiveFile = inactiveFile;
    this.chain = chain;

    this.addresses = this.loadAddresses();
    this.processed = 0;
    this.active = 0;
    this.inactive = 0;
    this.retries = 0;

    this.rpcManager = this.buildRpcManager();
    this.checker = new CheckerClass(this.rpcManager);
  }

  loadAddresses() {
    try {
      return fs
        .readFileSync(this.inputFile, "utf8")
        .split(/\r?\n/)
        .map((line) => line.trim())
        .filter(Boolean);
    } catch {
      return [];
    }
  }

  buildRpcManager() {
    const entries = [];

    try {
      const lines = fs.readFileSync(this.rpcFile, "utf8").split(/\r?\n/);

      for (const rawLine of lines) {
        const line = rawLine.trim();
        if (!line) {
          continue;
        }

        if (line.includes("|")) {
          const [leftRaw, rightRaw] = line.split("|", 2);
          const left = leftRaw.trim();
          const right = rightRaw.trim();

          if (left.startsWith("http://") || left.startsWith("https://")) {
            entries.push({ mode: right, url: left });
          } else {
            entries.push({ mode: left, url: right });
          }
        } else {
          entries.push({ mode: "RPC", url: line });
        }
      }
    } catch {
      return new RPCManager([]);
    }

    return new RPCManager(entries);
  }

  async worker(address) {
    try {
      const result = await this.checker.check(address);
      this.processed += 1;

      if (result) {
        this.active += 1;
        fs.mkdirSync(path.dirname(this.activeFile), { recursive: true });
        fs.appendFileSync(this.activeFile, `${this.chain},${address}\n`);
      } else {
        this.inactive += 1;
      }
    } catch {
      this.retries += 1;
    }
  }

  async run() {
    let tasks = [];

    for (const address of this.addresses) {
      tasks.push(this.worker(address));

      if (tasks.length >= 100) {
        await Promise.all(tasks);
        tasks = [];
      }
    }

    if (tasks.length > 0) {
      await Promise.all(tasks);
    }
  }

  reset() {
    this.addresses = this.loadAddresses();
    this.processed = 0;
    this.active = 0;
    this.inactive = 0;
    this.retries = 0;
  }

  status() {
    return {
      chain: this.chain,
      processed: this.processed,
      active: this.active,
      inactive: this.inactive,
      retries: this.retries,
      rpcStatus: this.rpcManager.rpcs.map((entry, index) => ({
        url: entry.url,
        state: (this.rpcManager.failCount.get(index) || 0) === 0 ? "available" : "retrying"
      }))
    };
  }
}

module.exports = { Engine };